airbyte-integrations/connectors/source-surveymonkey/source_surveymonkey/streams.py

```diff
@@ -0,0 +1,265 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+import urllib.parse
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional
+
+import pendulum
+import requests
+import vcr
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.http import HttpStream
+
+
+class SurveymonkeyStream(HttpStream, ABC):
+    url_base = "https://api.surveymonkey.com/v3/"
+    primary_key = "id"
+    data_field = "data"
+
+    def __init__(self, start_date: str, **kwargs):
+        if isinstance(start_date, str):
+            start_date = pendulum.parse(start_date)  # convert to YYYY-MM-DDTHH:MM:SS
+        self._start_date = start_date
+        super().__init__(**kwargs)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        links = resp_json.get("links", {})
+        if links.get("next"):
+            next_query_string = urllib.parse.urlsplit(links["next"]).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            return params
+
+    def request_headers(self, **kwargs) -> Mapping[str, Any]:
+        return {"Content-Type": "application/json"}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        if response_json.get("error"):
+            raise Exception(repr(response_json.get("error")))  # TODO: apply ShouldRetry
+        result = response.json.get(self.data_field, [])
+        yield from result
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {}
+        if next_page_token:
+            params.update(next_page_token)
+        return params
+
+    def read_records(  # TODO remove this (caching) in final commit
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        with vcr.use_cassette("pretty_name.yaml", record_mode="new_episodes"):
```

Comment:
Why do you use pretty_name.yaml file?

Reply:
I think it's where cached data file will be placed.

Reply:
yaml is a output file format for vcr library - it uses it for caching responses. The limit will be reached very fast (500 requests is only 3 test runs) if not to do that, we have very limited api.